packages/dev/core/src/AudioV2/abstractAudio/subNodes/abstractAudioSubGraph.ts

```diff
@@ -0,0 +1,146 @@
+     *
+     * @param name The name of the sub node
+     * @param callback The function to call with the named sub node
+     */
+    public callOnSubNode<T extends _AbstractAudioSubNode>(name: string, callback: (node: T) => void): void {
+        const node = this.getSubNode(name);
+        if (node) {
+            callback(node as T);
+            return;
+        }
+
+        const promise = this._createSubNodePromises[name] ?? this._createAndAddSubNode(name);
+
+        promise?.then((node) => {
+            callback(node as T);
+        });
+    }
+
+    /**
+     * Releases associated resources.
+     */
+    public dispose() {
+        const subNodeSets = Object.values(this._subNodes);
+        for (const subNodeSet of subNodeSets) {
+            const subNodeSetIterator = subNodeSet.values();
+            for (let nextSubNodeSet = subNodeSetIterator.next(); !nextSubNodeSet.done; nextSubNodeSet = subNodeSetIterator.next()) {
+                nextSubNodeSet.value.dispose();
+            }
+
+            subNodeSet.clear();
+        }
+
+        this._subNodes = {};
+        this._createSubNodePromises = {};
+    }
+
+    /**
+     * Gets a previously created sub node.
+     * @param name - The name of the sub node
+     * @returns The named sub node, or `null` if has not been created, yet
+     * @internal
+     * */
+    public getSubNode<T extends AbstractNamedAudioNode>(name: string): Nullable<T> {
+        const set = this._subNodes[name];
+
+        if (!set) {
+            return null;
+        }
+
+        return set.values().next().value as T;
+    }
+
+    protected abstract _createSubNode(name: string): Nullable<Promise<_AbstractAudioSubNode>>;
+
+    /**
+     * Called when sub-nodes are added or removed.
+     * - Override this to connect and reconnect sub-nodes as needed.
+     */
+    protected _onSubNodesChanged(): void {}
+
+    protected _createSubNodePromisesResolved(): Promise<_AbstractAudioSubNode[]> {
+        return Promise.all(Object.values(this._createSubNodePromises));
+    }
+
+    private _getSubNodeSet(name: string): Set<AbstractNamedAudioNode> {
+        let set = this._subNodes[name];
+
+        if (!set) {
+            set = new Set<AbstractNamedAudioNode>();
+            this._subNodes[name] = set;
+        }
+
+        return set;
+    }
+
+    private _addSubNode(node: AbstractNamedAudioNode): void {
+        this._getSubNodeSet(node.name).add(node);
+        this._onSubNodesChanged();
+
+        node.onDisposeObservable.addOnce(this._onSubNodeDisposed);
+        node.onNameChangedObservable.add(this._onSubNodeNameChanged);
+    }
+
+    protected _createAndAddSubNode(name: string): Nullable<Promise<_AbstractAudioSubNode>> {
+        const promise = this._createSubNode(name);
+
+        if (!promise) {
+            return null;
+        }
+
+        promise.then((node) => {
+            this._addSubNode(node);
+        });
+
+        this._createSubNodePromises[name] = promise;
```

Comment:
Don't you want this to be resolved after `this._addSubNode` is called? Currently, it will resolve before.

Reply:
Yes. Should be fixed in commit https://github.com/BabylonJS/Babylon.js/pull/15839/commits/c6f989bf0c48be5f55fba57fb4f6290ede0462a5.